### PR TITLE
Extract string literals to constants, as they are prone to typos

### PR DESCRIPTION
**Pull request by Anishka0107**

### CODE REVIEW

**Comment by Anishka0107** [comment]
I have to do the same for a few other files (basically the code & new files added in PR #74).

File: src/vcf/record.cpp [hunk 1]
```diff
@@ -344,33 +344,33 @@ namespace ebi
...
+            auto it = info.find(IMPRECISE);
             if (it != info.end() && it->second == "0") {
                 auto expected = std::to_string(position + reference_allele.length() - 1);
                 if (field_value != expected) {
                     throw new InfoBodyError{line, "INFO END=" + field_value + " value must be equal to \"POS + length of REF - 1\" for a precise variant (where IMPRECISE is set to 0), expected " + expected, field_key};
                 }
             }
-        } else if (field_key == "SVLEN" && values.size() == alternate_alleles.size()) {
+        } else if (field_key == SVLEN && values.size() == alternate_alleles.size()) {
             bool is_ref_symbolic = true;
```
Comment (jmmut) [inline src/vcf/record.cpp]: this is used sometimes properly and sometimes as `is_ref_not_symbolic`. later in this function, the code under `if (is_ref_symbolic && is_alt_symbolic[i])` is in fact what we should do if the alleles are not symbolic, which is taking the difference in allele string sizes. also, the reference can not be a symbolic allele (the ragel code checks that under the record_ref machine)

File: src/vcf/record.cpp [hunk 2]
```diff
@@ -344,33 +344,33 @@ namespace ebi
...
                 if (!boost::regex_match(value, cigar_string)) {
                     throw new InfoBodyError{line, "INFO CIGAR=" + field_value + " value is not an alphanumeric string compliant with the SAM specification", field_key};
                 }
             }
-        } else if (field_key == "END") {
-            auto it = info.find("IMPRECISE");
+        } else if (field_key == END) {
+            auto it = info.find(IMPRECISE);
             if (it != info.end() && it->second == "0") {
                 auto expected = std::to_string(position + reference_allele.length() - 1);
```
Comment (jmmut) [inline src/vcf/record.cpp]: I think I had an oversight here, shouldn't this be done only if reference is not symbolic?

File: test/vcf/predefined_format_tags_test.cpp [hunk 3]
```diff
@@ -42,12 +42,12 @@ namespace ebi
                             "chr1",
                             123456,
                             { "id123" },
-                            "A",
+                            vcf::A,
```
Comment (jmmut) [inline test/vcf/predefined_format_tags_test.cpp]: This usages (bases string) should not be extracted as the same constant for "cardinality A". revert all usages in this file, please

File: test/vcf/predefined_info_tags_test.cpp [hunk 4]
```diff
@@ -42,12 +42,12 @@ namespace ebi
                             "chr1",
                             123456,
                             { "id123" },
-                            "A",
+                            vcf::A,
```
Comment (jmmut) [inline test/vcf/predefined_info_tags_test.cpp]: same as predefined_format_tags_test, this usages (bases string) should not be extracted as the same constant for "cardinality A". revert all usages in this file, please

File: test/vcf/record_test.cpp [hunk 5]
```diff
@@ -35,54 +35,54 @@ namespace ebi
...
                 1,
-                "INFO",
+                vcf::INFO,
                 {
-                    { "ID", "AF" },
-                    { "Number", "A" },
-                    { "Type", "Float" },
-                    { "Description", "Allele frequency" }
+                    { vcf::ID, vcf::AF },
+                    { vcf::NUMBER, vcf::A },
```
Comment (jmmut) [inline test/vcf/record_test.cpp]: except for this usage, which is ok, revert all the other usages of "A" in this file as well

File: test/vcf/record_test.cpp [hunk 6]
```diff
@@ -95,12 +95,12 @@ namespace ebi
                                 "chr1", 
                                 123456, 
                                 { "id123", "id456" }, 
-                                "A", 
-                                { "AC", "AT" }, 
+                                vcf::A, 
+                                { vcf::AC, "AT" },
```
Comment (jmmut) [inline test/vcf/record_test.cpp]: revert back AC to hardcoded string in this file, as it is used as a bases string too.

File: src/vcf/record.cpp (hunk 2 above)
Comment (jmmut) [inline src/vcf/record.cpp]: nevermind, remembered that references cannot be symbolic

File: src/vcf/meta_entry.cpp [hunk 7]
```diff
@@ -83,172 +83,172 @@ namespace ebi
...
         if (iterator != tags.end()) {
             // Determine the required value of the key based on whether we are checking for Type or Number
-            std::string key_value = (key_field == "Type" ? iterator->second.first : iterator->second.second);
+            std::string key_value = (key_field == TYPE ? iterator->second.first : iterator->second.second);
             // If the required value is a "." (dot), do nothing
             // Or if the required value does not match the value provided in the vcf file, throw an error
-            if (key_value != "." && key_value != value[key_field]) {
-                std::string message = tag_field + " " + value["ID"] + " metadata " + key_field + " is not " + key_value;
+            if (key_value != MISSING_VALUE && key_value != value[key_field]) {
+                std::string message = tag_field + " " + value[ID] + " metadata " + key_field + " is not " + key_value;
```
Comment (Anishka0107) [inline src/vcf/meta_entry.cpp]: I agree they sound much better!

File: test/vcf/predefined_format_tags_test.cpp (hunk 3 above)
Comment (Anishka0107) [inline test/vcf/predefined_format_tags_test.cpp]: Oh sorry, I did a search & replace, so got this wrong

File: test/vcf/record_test.cpp [hunk 8]
```diff
@@ -481,7 +481,7 @@ namespace ebi
                 vcf::INFO,
                 {
                     { vcf::ID, vcf::AF },
-                    { vcf::NUMBER, vcf::A },
+                    { vcf::NUMBER, "A" },
```
Comment (jmmut) [inline test/vcf/record_test.cpp]: oops, this usage was correct as vcf::A, I thought there was only one usage of this at the beginning of this file

File: test/vcf/record_test.cpp [hunk 9]
```diff
@@ -110,11 +110,11 @@ namespace ebi
                                 123456, 
                                 { }, 
                                 "A", 
-                                { "AC", "AT" }, 
+                                { vcf::AC, "AT" },
```
Comment (jmmut) [inline test/vcf/record_test.cpp]: all the usages of vcf::AC in this file should be "AC" actually

File: src/vcf/record.cpp [hunk 10]
```diff
@@ -416,6 +392,19 @@ namespace ebi
         }
     }
 
+    bool Record::check_alt_not_symbolic(size_t i) const
```
Comment (cyenyxe) [inline src/vcf/record.cpp]: Renaming `i` to `allele_index` would clarify its purpose.

File: inc/vcf/string_constants.hpp [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+  namespace vcf
+  {
+
+    // VCF versions
+    const std::string VCF_V41 = "VCFv4.1";
+    const std::string VCF_V42 = "VCFv4.2";
+    const std::string VCF_V43 = "VCFv4.3";
+
+    // validator and debugulator strings
+    const char STDIN[] = "stdin";
```
Comment (cyenyxe) [inline inc/vcf/string_constants.hpp]: Does boost require the arguments to be `const char*` instead of `string`?

Comment (Anishka0107) [inline inc/vcf/string_constants.hpp]: Yes

**Pull request merged**